File: Suspension matlab/spring/Time_sensitivity_analysis/Time_sensitivity_analysis.m
%% Explicit Euler
% Time Step Sensitivity Analysis
clear; clc; close all;
scriptInfo = dbstack('-completenames');
scriptDir = fileparts(scriptInfo(1).file);
addpath(scriptDir);
params = load_data(fullfile(scriptDir,'data.csv'));
paramNames = fieldnames(params);
for paramIdx = 1:numel(paramNames)
    eval([paramNames{paramIdx} ' = params.(paramNames{paramIdx});']);
end

%% Parameters

m = time_m;      % kg
k = time_k;      % N/m

%% Initial conditions

x0 = time_x0;      % m
v0 = time_v0;      % m/s

%% Time settings

t_end = time_t_end;

dt_list = time_dt_list;

%% Loop for different dt

for d = 1:length(dt_list)

    dt = dt_list(d);

    t = 0:dt:t_end;
    N = length(t);

    %% Initialize

    x = zeros(1,N);
    v = zeros(1,N);
    a = zeros(1,N);

    x(1) = x0;
    v(1) = v0;

    %% Explicit Euler

    for i = 1:N-1

        % acceleration
        a(i) = -(k/m)*x(i);

        % velocity update
        v(i+1) = v(i) + a(i)*dt;

        % position update
        x(i+1) = x(i) + v(i)*dt;

    end

    %% Final acceleration

    a(N) = -(k/m)*x(N);

    %% Energy calculation

    KE = 0.5 * m .* v.^2;
    PE = 0.5 * k .* x.^2;

    E_total = KE + PE;

    E0 = E_total(1);

    %% Energy error %

    E_error_percent = ...
        (E_total - E0)/E0 * 100;

    %% =====================================
    % Plot
    %% =====================================

    figure('Name',['dt = ' num2str(dt)]);

    %% Displacement

    subplot(3,1,1)

    plot(t,x,'LineWidth',2);

    grid on;

    ylabel('Displacement (m)');

    title(['Explicit Euler  |  dt = ' num2str(dt)]);

    %% Total Energy

    subplot(3,1,2)

    plot(t,E_total,'LineWidth',2);

    grid on;

    ylabel('Energy (J)');

    %% Energy Error

    subplot(3,1,3)

    plot(t,E_error_percent,'LineWidth',2);

    grid on;

    xlabel('Time (s)');
    ylabel('Energy Error (%)');

end

File: Suspension matlab/spring/Time_sensitivity_analysis/Time_sensitivity_analysis.py
from pathlib import Path
import sys

import matplotlib.pyplot as plt
import numpy as np

ROOT = next(p for p in Path(__file__).resolve().parents if (p / "suspension_utils.py").exists())
sys.path.insert(0, str(ROOT))

from suspension_utils import load_data, show_plots


def main() -> None:
    p = load_data(Path(__file__).with_name("data.csv"))
    m = p["time_m"]
    k = p["time_k"]
    x0 = p["time_x0"]
    v0 = p["time_v0"]
    t_end = p["time_t_end"]
    dt_list = np.asarray(p["time_dt_list"], dtype=float)

    for dt in dt_list:
        t = np.arange(0, t_end + dt / 2, dt)
        x = np.zeros_like(t)
        v = np.zeros_like(t)
        a = np.zeros_like(t)
        x[0] = x0
        v[0] = v0

        for i in range(len(t) - 1):
            a[i] = -(k / m) * x[i]
            v[i + 1] = v[i] + a[i] * dt
            x[i + 1] = x[i] + v[i] * dt
        a[-1] = -(k / m) * x[-1]

        ke = 0.5 * m * v**2
        pe = 0.5 * k * x**2
        e_total = ke + pe
        e_error_percent = (e_total - e_total[0]) / e_total[0] * 100

        fig, axes = plt.subplots(3, 1, num=f"dt = {dt:g}")
        axes[0].plot(t, x, linewidth=2)
        axes[0].grid(True)
        axes[0].set_ylabel("Displacement (m)")
        axes[0].set_title(f"Explicit Euler  |  dt = {dt:g}")
        axes[1].plot(t, e_total, linewidth=2)
        axes[1].grid(True)
        axes[1].set_ylabel("Energy (J)")
        axes[2].plot(t, e_error_percent, linewidth=2)
        axes[2].grid(True)
        axes[2].set_xlabel("Time (s)")
        axes[2].set_ylabel("Energy Error (%)")
    show_plots()


if __name__ == "__main__":
    main()
